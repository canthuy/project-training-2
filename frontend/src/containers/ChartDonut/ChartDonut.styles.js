import styled from 'styled-components';

const ChartSC = styled.div`
  width: 30%;
  margin: 20px;
`;

export { ChartSC };
