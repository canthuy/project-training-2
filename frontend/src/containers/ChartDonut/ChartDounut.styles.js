import styled from 'styled-components';

const ChartSC = styled.div`
  width: 30%;
  margin: 10px auto;
`;

const Error = styled.h2`
  text-align: center;
  color: #be0f0f;
  margin: 20px;
`;
export { ChartSC, Error };
