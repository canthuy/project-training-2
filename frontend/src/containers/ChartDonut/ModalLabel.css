.modal-custom {
  width: 30% !important;
  position: absolute !important;
  top: 5rem;
  left: 1rem;
}
.input-filter {
  width: 100%;
  border: none;
  outline: none;
  border-bottom: 3px solid #3cb4ec;
}

.btn-labels:active, .btn-labels:focus {
    box-shadow: none !important;
}
.label-os {
    padding: 10px 10px 0;
    display: flex;
    justify-content: space-between;
}
.label-os span {
    font-size: 16px
}
.label-os:hover{
    cursor: pointer;
    color: #3cb4ec
}