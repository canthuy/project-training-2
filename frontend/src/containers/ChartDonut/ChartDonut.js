import { Chart, ArcElement, Tooltip, Legend, Title } from 'chart.js';
import { Doughnut } from 'react-chartjs-2';
import { useSelector, useDispatch } from 'react-redux';
import { Fragment, useEffect } from 'react';
import { getData } from '../../redux/doughnut/actions';
import { ChartSC } from './ChartDonut.styles';
import Loading from '../../components/Loading/Loading';
import Error from '../../components/Error/Error';
import ModalLabel from './ModalLabel';
Chart.register(ArcElement, Tooltip, Legend, Title);

const ChartDoughnut = () => {
  const { deviceData, loading, isError } = useSelector((state) => state.device);
  const { startDate, endDate } = useSelector((state) => state.datepicker);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getData([startDate, endDate]));
  }, [startDate, endDate]);
  const options = {
    plugins: {
      legend: {
        display: true,
        position: 'right',
        labels: {
          boxWidth: 15,
          boxHeight: 10,
          padding: 20,
          font: {
            size: 16,
          },
        },
      },
      title: {
        display: true,
        text: 'Decive Type',
        align: 'start',
        font: {
          size: 18,
          weight: 0,
        },
        padding: {
          top: 20,
        },
      },
    },
    elements: {
      arc: {
        borderWidth: 0,
      },
    },
    cutout: '75%',
  };
  const data = {
    labels: deviceData.map((val) => {
      return val.x;
    }),
    datasets: [
      {
        data: deviceData.map((val) => {
          return val.y;
        }),
        backgroundColor: [
          '#48c0b0',
          '#925de2',
          '#36a2eb',
          '#c9cbcf',
          '#ffcd56',
          '#f77b00',
        ],
        hoverBackgroundColor: [
          '#48c0b0',
          '#925de2',
          '#36a2eb',
          '#c9cbcf',
          '#ffcd56',
          '#f77b00',
        ],
      },
    ],
  };
  const displayChart = (
    <ChartSC>
      <ModalLabel />
      <Doughnut data={data} options={options} />
    </ChartSC>
  );
  return (
    <Fragment>
      {loading && (
        <ChartSC>
          <Loading />
        </ChartSC>
      )}
      {isError && <Error message="Error: Network Error" />}
      {!isError && !loading && displayChart}
    </Fragment>
  );
};
export default ChartDoughnut;
