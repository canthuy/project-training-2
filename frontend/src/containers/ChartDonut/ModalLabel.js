import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getData } from '../../redux/doughnut/actions';

import { Modal, Button } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck } from '@fortawesome/free-solid-svg-icons';

import './ModalLabel.css';

const ModalLabel = () => {
  const listDevices = ['Android', 'iOS', 'Windows', 'Os X', 'Unknown', 'Linux'];
  const [listLabel, setListLabel] = useState(listDevices);
  const [show, setShow] = useState(false);
  const deviceData = useSelector((state) => state.device.deviceData);
  const { startDate, endDate } = useSelector((state) => state.datepicker);
  const [listChecked, setListChecked] = useState(
    deviceData.map((val) => val.x)
  );
  const dispatch = useDispatch();

  const onClickLabel = (val) => {
    setListChecked((prevList) => {
      const isChecked = prevList.includes(val);
      if (isChecked) {
        return prevList.filter((check) => check !== val);
      }
      return [...prevList, val];
    });
  };

  const onHideModal = () => {
    const prevListChecked = deviceData.map((val) => val.x);
    const isTrue =
      listChecked.length === prevListChecked.length &&
      prevListChecked.every((v) => listChecked.includes(v));

    if (listChecked.length > 0) {
      if (!isTrue) {
        dispatch(getData([startDate, endDate, listChecked]));
      } else {
        setShow(false);
      }
    } else {
      alert('Please select device type');
    }
  };

  const handleFilter = (e) => {
    e.preventDefault();
    const keyFilter = e.target.value.trim();
    if (keyFilter.length > 0) {
      setListLabel((prev) => {
        return prev.filter((val) => val.toLowerCase().includes(keyFilter));
      });
    } else {
      setListLabel(listDevices);
    }
  };

  return (
    <>
      <Button className="btn-labels" size="sm" onClick={() => setShow(true)}>
        Labels
      </Button>
      <Modal
        show={show}
        onHide={onHideModal}
        dialogClassName="modal-custom"
        aria-labelledby="example-custom-modal-styling-title"
      >
        <Modal.Body>
          <input className="input-filter" onChange={handleFilter} />
          {listLabel.map((val, index) => {
            return (
              <div
                key={index}
                className="label-os"
                onClick={() => onClickLabel(val)}
              >
                <span>{val}</span>
                {listChecked.includes(val) && (
                  <FontAwesomeIcon icon={faCheck} data-testid="test-svg" />
                )}
              </div>
            );
          })}
        </Modal.Body>
      </Modal>
    </>
  );
};
export default ModalLabel;
