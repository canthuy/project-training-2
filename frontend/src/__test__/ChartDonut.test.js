import React from 'react';
import { render, screen } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import renderer from 'react-test-renderer';
import ChartDonut from '../containers/ChartDonut/ChartDonut';
window.React = React;

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('ChartDonut Component', () => {
  let mockAppState = {};
  beforeEach(() => {
    useDispatch.mockImplementation(() => () => {});
    useSelector.mockImplementation((callback) => {
      return callback(mockAppState);
    });
  });
  afterEach(() => {
    useSelector.mockClear();
    useDispatch.mockClear();
  });
  it('loading data', ()=>{
    mockAppState = {
      deviceData: {},
      loading: true,
      isError: false,
    };
    const { container } = render(<ChartDonut />);
    const elements = container.querySelector('div');
    const tree = renderer.create(<ChartDonut />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(elements.firstChild.className).toEqual('spinner-border');
  });
  it('get data success', () => {
    mockAppState = {
      deviceData: { android: 50, iOS: 50 },
      loading: false,
      isError: false,
    };
    const { container } = render(<ChartDonut />);
    const tree = renderer.create(<ChartDonut />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(container.querySelector('canvas')).toBeTruthy();  
  });
  it('get data error', () => {
    mockAppState = {
      deviceData: {},
      loading: false,
      isError: true,
    };
    render(<ChartDonut />);
    const tree = renderer.create(<ChartDonut />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(screen.getByText('Error: Network Error')).toBeTruthy();
  });
});
