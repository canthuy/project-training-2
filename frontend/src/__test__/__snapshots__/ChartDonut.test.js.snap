// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartDonut Component get data error 1`] = `
<div>
  <h2
    class="sc-dkPtRN cZRHky"
  >
    Error: Network Error
  </h2>
</div>
`;

exports[`ChartDonut Component get data success 1`] = `
<div>
  <div
    class="sc-bdvvtL cfVjDS"
  >
    <canvas
      height="150"
      role="img"
      width="300"
    />
  </div>
</div>
`;

exports[`ChartDonut Component loading data 1`] = `
<div>
  <div
    class="sc-bdvvtL cfVjDS"
  >
    <div
      class="sc-gsDKAQ giLnYi"
    >
      <div
        class="spinner-border"
      />
    </div>
  </div>
</div>
`;
